src/coreneuron/io/core2nrn_data_return.cpp

```diff
@@ -179,6 +184,62 @@ static void core2nrn_corepointer(int tid, NrnThreadMembList* tml) {
     (*core2nrn_corepointer_mech_)(tid, type, icnt, dcnt, iArray.get(), dArray.get());
 }
 
+// based on code from nrncore_callbacks.cpp
+static std::unordered_map<int, std::vector<int>> mech_random_indices{};
+std::vector<int>& nrn_mech_random_indices(int type) {
+    // not threadsafe until after first call for a type
```

Comment:
Presently, both uses (transfer back to NEURON and checkpointing) are serial. But I think it would be a good idea to lock (can one lock just an if statement?)

Reply:
I think something like that would do?

```c++
static std::unordered_map<int, std::vector<int>> mech_random_indices{};

std::vector<int>& nrn_mech_random_indices(int type) {
    static std::mutex mx;
    std::unique_lock<std::mutex> lock(mx);
    if (mech_random_indices.count(type) == 0) {
        // if no element, create empty one and search dparam_semantics to fill
        auto& mri = mech_random_indices[type];
        int* semantics = corenrn.get_memb_func(type).dparam_semantics;
        int dparam_size = corenrn.get_prop_dparam_size()[type];
        for (int i = 0; i < dparam_size; ++i) {
            if (semantics[i] == -11) {
                mri.push_back(i);
            }
        }
    }
    lock.unlock();
    return mech_random_indices[type];
}
```

but my c++ multithreading knowledge is a bit vague I have to admit - maybe @1uc could comment on whether I just introduced a bug with this or not :)

Reply:
This does not protect against one thread reading while another is writing. It doesn't protect/can't when one thread is writing via `nrn_mech_random_indices` and another is writing directly to the global variable.

Reply:
Thanks luc - I thought that reading while writing is ok - but you're probably right that this would be also problematic. But more generally, @nrnhines after some more discussion we had in our team, I think we can just drop this and keep it "unsafe" as it is. The current implementation is anyway not multi threaded and if it would be in the future made multi threaded then we would have many other issues that would need to addressed - so it seems a bit a waste to block this PR for this.

Reply:
You weren't imprecise, I just passed over the statement ```and another is writing directly to the global variable.``` without comprehending its import.   I'll move the definition of ```mech_random_indices``` to be a static variable inside ```nrn_mech_random_indices```.